pkg/controllers/util/listwatcher/listwatcher.go

```diff
@@ -0,0 +1,239 @@
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.receivers = append(p.receivers, receiver)
+}
+
+func (p *pagedListWatcher) Start(ctx context.Context) {
+	logger := klog.FromContext(ctx).WithValues("component", "paged-list-watcher")
+	listCh := make(chan string, 1)
+
+	go func() {
+		for {
+			// list resource
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				p.mux <- Event{Type: StartSync}
+
+				resourceVersion, err := p.list(ctx)
+				if err != nil {
+					logger.Error(err, "Failed to list resource, will retry after backoff")
+					<-p.backoff.Backoff().C()
+					continue
+				}
+
+				p.mux <- Event{Type: EndSync}
+				listCh <- resourceVersion
+			}
+
+			// watch resource
+			select {
+			case <-ctx.Done():
+				return
+			case resourceVersion := <-listCh:
+				if err := p.watch(ctx, resourceVersion); err != nil {
+					logger.Error(err, "Watch resource ended with error, will relist after backoff")
+					<-p.backoff.Backoff().C()
+				} else {
+					logger.V(4).Info("Watch resource ended gracefully, will relist")
+				}
+			}
+		}
+	}()
+
+	go func() {
+		for event := range p.mux {
+			p.mu.Lock()
+			for _, receiver := range p.receivers {
+				receiver <- event
+			}
+			p.mu.Unlock()
+		}
+	}()
+}
+
+func (p *pagedListWatcher) list(ctx context.Context) (rv string, err error) {
+	var resourceVersion string
+	listOpts := metav1.ListOptions{
+		Limit:    p.pageSize,
+		Continue: "",
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return "", nil
+		default:
+			resp, err := p.lw.List(listOpts)
+			if err != nil {
+				return "", fmt.Errorf("failed to list resource: %w", err)
+			}
+
+			list, err := meta.ListAccessor(resp)
+			if err != nil {
+				return "", fmt.Errorf("failed to list resource: %w", err)
+			}
+
+			listOpts.Continue = list.GetContinue()
+			resourceVersion = list.GetResourceVersion()
+
+			err = meta.EachListItem(resp, func(o runtime.Object) error {
+				p.mux <- Event{
+					Type:   Add,
+					Object: o.DeepCopyObject(),
+				}
+				return nil
+			})
+
+			if err != nil {
+				return "", fmt.Errorf("failed to process list item: %w", err)
+			}
+		}
+
+		if len(listOpts.Continue) == 0 {
+			break
+		}
+	}
+
+	return resourceVersion, nil
+}
+
+func (p *pagedListWatcher) watch(ctx context.Context, resourceVersion string) error {
+	watcher, err := p.lw.Watch(metav1.ListOptions{ResourceVersion: resourceVersion})
+	if err != nil {
+		return fmt.Errorf("failed to watch resource: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				return nil
+			}
+
+			switch event.Type {
+			case watch.Added:
+				p.mux <- Event{
+					Type:   Add,
+					Object: event.Object,
+				}
+			case watch.Modified:
+				p.mux <- Event{
+					Type:   Update,
+					Object: event.Object,
+				}
+			case watch.Deleted:
+				p.mux <- Event{
+					Type:   Delete,
+					Object: event.Object,
+				}
+			case watch.Error:
```

Comment:
https://github.com/kubernetes/client-go/blob/8cbca742aebe24b24f7f4e32fd999942fa9133e8/tools/cache/reflector.go#L332

Do we need to handle these watch errors?
```go
			if err != errorStopRequested {
				switch {
				case isExpiredError(err):
					// Don't set LastSyncResourceVersionUnavailable - LIST call with ResourceVersion=RV already
					// has a semantic that it returns data at least as fresh as provided RV.
					// So first try to LIST with setting RV to resource version of last observed object.
					klog.V(4).Infof("%s: watch of %v closed with: %v", r.name, r.expectedTypeName, err)
				case apierrors.IsTooManyRequests(err):
					klog.V(2).Infof("%s: watch of %v returned 429 - backing off", r.name, r.expectedTypeName)
					<-r.initConnBackoffManager.Backoff().C()
					continue
				case apierrors.IsInternalError(err) && retry.ShouldRetry():
					klog.V(2).Infof("%s: retrying watch of %v internal error: %v", r.name, r.expectedTypeName, err)
					continue
				default:
					klog.Warningf("%s: watch of %v ended with: %v", r.name, r.expectedTypeName, err)
				}
			}
```